### PR TITLE
Avoid concurrent initdb calls for a timeline

### PR DESCRIPTION
**Pull request by SomeoneToIgnore**
Takes a lock before running initdband check initdb target dir for existence, to avoid concurrent runs.
Currently, related lock manipulations are in tenant.rs, so the timeline creation code needs to be moved there from `timelines.rs`

### CODE REVIEW

File: pageserver/src/tenant.rs [hunk 1]
```diff
@@ -952,11 +1019,95 @@ impl Tenant {
...
+    /// - after initialization complete, remove the temp dir.
+    async fn bootstrap_timeline(
+        &self,
+        conf: &'static PageServerConf,
+        timeline_id: TimelineId,
+    ) -> Result<Arc<Timeline>> {
+        // XXX: ensure no other condurrent processes run initdb for the new timeline
+        let timelines_guard = self.timelines.lock();
+
+        // construct a `tenant/{tenant_id}/timelines/basebackup-{timeline_id}.{TEMP_FILE_SUFFIX}/`
```
Comment (hlinnaka) [inline pageserver/src/tenant.rs:1032]:
```suggestion
        // construct a `tenants/{tenant_id}/timelines/basebackup-{timeline_id}.{TEMP_FILE_SUFFIX}/`
```

File: pageserver/src/tenant.rs [hunk 2]
```diff
@@ -952,11 +1019,95 @@ impl Tenant {
...
-        self.tenant_id
+    /// - run initdb to init temporary instance and get bootstrap data
+    /// - after initialization complete, remove the temp dir.
+    async fn bootstrap_timeline(
+        &self,
+        conf: &'static PageServerConf,
+        timeline_id: TimelineId,
+    ) -> Result<Arc<Timeline>> {
+        // XXX: ensure no other condurrent processes run initdb for the new timeline
+        let timelines_guard = self.timelines.lock();
```
Comment (hlinnaka) [inline pageserver/src/tenant.rs:1030]: Instead of grabbing the lock, I'd suggest running initdb in a known unique path. So instead of `.temp` suffix, something like `.<random string>.temp`. That way, there cannot be a conflict while running initdb.

Comment (SomeoneToIgnore) [reply pageserver/src/tenant.rs:1030]: That is a bit complicated, because it hardens the tmp dir check on restart: we would need to somehow detect such directories for the timeline and then to somehow verify that the timeline was initialized.

With the PR proposed model, if we find such directory near the timeline, we know that the timeline was not initialized correctly (due to that tmp dir being last) hence both tmp dir and the timeline need purging.

Yet with multiple such directories, we can import initdb result from one of the runs and fail the others with pageserver restart so, that there are some tmp initdb dirs and a properly initialized timeline.
I would imagine, it's ok to check `disk_consistent_lsn` in such case to ensure it's on 0 and it should be enough, but it's rather subtle.

Comment (hlinnaka) [reply pageserver/src/tenant.rs:1030]: > With the PR proposed model, if we find such directory near the timeline, we know that the timeline was not initialized correctly (due to that tmp dir being last) hence both tmp dir and the timeline need purging.

Oh, that's very subtle! So if there are two directories at pageserver startup:
```
tenants/1234/timelines/5678.temp
tenants/1234/timelines/5678
```
It will remove *both*?

Ok, yes, that works. Needs a big comment somewhere! And makes me a bit uncomfortable; it seems safe to delete anything with the `.temp` suffix, but it feels much more dangerous to delete something that looks valid, just because a different directory with `.temp` suffix exists.

IMHO a nicer way to do this would be to create the directory for the new, empty, timeline with `.temp` suffix, import all the data there, and finally once it's ready, rename it without the .temp suffix. That would require more code changes though. Currently we assume in many places that the path for a timeline is always `tenants/<tenant_id>/timelines/<timeline_id>`, and now during import it would be `tenants/<tenant_id>/timelines/<timeline_id>.temp`.

In summary, I'm OK with this PR for now, but needs a comment somewhere, and in the long run I think we should consider that other approach.

Comment (SomeoneToIgnore) [reply pageserver/src/tenant.rs:1030]: > It will remove both?

Not right now, I've planned to add that in another PR, but will add here, since apparently very related.
But yes, that's the intention and that's how we're sometimes left with 0-init "broken" timelines after load tests, presumably: I see no other way but to remove both directories with the current code approach.

@LizardWizzard and I discussed tombstones a bit as a solution to the issue, but it all looks quite tangled now, so whatever we try, it will require plenty of changes.
Not happy with lock acquisition either, but see no other viable options.

I wonder, how appropriate it is for such operations to go and create the tmp directory right away, without any locks?
If all concurrent operations fail (aborting tenant/timeline/whatever creation) and only one creates the directory and proceeds with init, we get ourselves a "good" race without extra locking?

**Comment by SomeoneToIgnore** [comment]
Had discussed it more offline, and looks like https://github.com/neondatabase/neon/pull/2451#discussion_r972033214 approach with directory creation races is indeed possible and looks better than current, proposed in this PR and https://github.com/neondatabase/neon/pull/2451#discussion_r972023678 alternatives to me, so closing this and trying the new one instead.
